Source/loadsave.cpp

```diff
@@ -882,369 +430,268 @@ void __cdecl SaveGame()
 }
 
 void __fastcall SavePremium(int i)
 {
-	memcpy(tbuff, &premiumitem[i], 0x170u);
-	tbuff = (char *)tbuff + 368;
+	memcpy(tbuff, &premiumitem[i], sizeof(*premiumitem));
+	tbuff += sizeof(*premiumitem);
 }
 
 void __fastcall SaveQuest(int i)
 {
-	memcpy(tbuff, &quests[i], 0x18u);
-	tbuff = (char *)tbuff + 24;
-	ISave(ReturnLvlX);
-	ISave(ReturnLvlY);
-	ISave(ReturnLvl);
-	ISave(ReturnLvlT);
-	ISave(DoomQuestState);
+	memcpy(tbuff, &quests[i], sizeof(*quests));
+	tbuff += sizeof(*quests);
+	WSave(ReturnLvlX);
+	WSave(ReturnLvlY);
+	WSave(ReturnLvl);
+	WSave(ReturnLvlT);
+	WSave(DoomQuestState);
 }
 
 void __fastcall SaveLighting(int i)
 {
-	memcpy(tbuff, &LightList[i], 0x34u);
-	tbuff = (char *)tbuff + 52;
+	memcpy(tbuff, &LightList[i], sizeof(*LightList));
+	tbuff += sizeof(*LightList);
 }
 
 void __fastcall SaveVision(int i)
 {
-	memcpy(tbuff, &VisionList[i], 0x34u);
-	tbuff = (char *)tbuff + 52;
+	memcpy(tbuff, &VisionList[i], sizeof(*VisionList));
+	tbuff += sizeof(*VisionList);
 }
 
 void __fastcall SavePortal(int i)
 {
-	memcpy(tbuff, &portal[i], 0x18u);
-	tbuff = (char *)tbuff + 24;
+	memcpy(tbuff, &portal[i], sizeof(*portal));
+	tbuff += sizeof(*portal);
 }
 
 void __cdecl SaveLevel()
 {
-	int v0; // eax
-	int i; // esi
-	int j; // esi
-	void *v47; // esi
-	int v48; // eax
-	int v49; // eax
-	char v50[260]; // [esp+0h] [ebp-10Ch]
-	void *SaveBuff; // [esp+104h] [ebp-8h]
-
-	if ( !currlevel )
+	int i, j;
+	char szName[MAX_PATH];
+
+	if(!currlevel)
 		glSeedTbl[0] = GetRndSeed();
-	v0 = codec_get_encoded_len(262147); /* FILEBUFF */
-	SaveBuff = DiabloAllocPtr(v0);
+
+	int dwLen = codec_get_encoded_len(FILEBUFF);
+	unsigned char *SaveBuff = DiabloAllocPtr(dwLen);
 	tbuff = SaveBuff;
-	if ( leveltype )
-	{
-		for(i = 0; i < 112; i++)
-		{
-			for(j = 0; j < 112; j++)
-			{
-				BSave(dDead[j][i]); /* check */
-			}
+
+	if(leveltype) {
+		for(j = 0; j < MAXDUNY; j++) {
+			for(i = 0; i < MAXDUNX; i++)
+				BSave(dDead[i][j]);
 		}
 	}
-	ISave(nummonsters);
-	ISave(numitems);
-	ISave(nobjects);
 
-	if ( leveltype )
-	{
-		for(i = 0; i < MAXMONSTERS; i++)
-			ISave(monstactive[i]);
+	WSave(nummonsters);
+	WSave(numitems);
+	WSave(nobjects);
 
+	if(leveltype) {
+		for(i = 0; i < MAXMONSTERS; i++)
+			WSave(monstactive[i]);
 		for(i = 0; i < nummonsters; i++)
 			SaveMonster(monstactive[i]);
-
 		for(i = 0; i < MAXOBJECTS; i++)
 			BSave(objectactive[i]);
-
 		for(i = 0; i < MAXOBJECTS; i++)
 			BSave(objectavail[i]);
-
 		for(i = 0; i < nobjects; i++)
 			SaveObject(objectactive[i]);
 	}
 
 	for(i = 0; i < MAXITEMS; i++)
 		BSave(itemactive[i]);
-
 	for(i = 0; i < MAXITEMS; i++)
 		BSave(itemavail[i]);
-
 	for(i = 0; i < numitems; i++)
 		SaveItem(itemactive[i]);
 
-	for(i = 0; i < 112; i++)
-	{
-		for(j = 0; j < 112; j++)
-		{
-			BSave(dFlags[j][i] & (DFLAG_POPULATED | DFLAG_MONSTER | DFLAG_PLAYER | DFLAG_LIT | DFLAG_EXPLORED));
-		}
+	for(j = 0; j < MAXDUNY; j++) {
+		for(i = 0; i < MAXDUNX; i++)
+			BSave(dFlags[i][j] & 0xF8);

```

Comment:
Lost the DFLAGs here

Reply:
Yeah, should be a not save. `~(FLAG1|FLAG...)`